spec/sinject/dependency_group_spec.rb

```diff
@@ -0,0 +1,74 @@
+# frozen_string_literal: true
+
+require 'spec_helper'
+require 'sinject/dependency_group'
+
+RSpec.describe Sinject::DependencyGroup do
+  describe '#register' do
+    it 'returns nil' do
+      expect(subject.register('anything')).to be nil
+    end
+  end
+
+  describe '.descendants' do
+    # Testing this method is awkward because it trawls ObjectSpace for
+    # DependencyGroup's own subclasses.
+    #
+    # To clean up after each test we therefore need to remove the
+    # definitions of any test DependencyGroup classes we've defined.
+    #
+    # The tests themselves need to be crafted carefully to ensure that
+    # Ruby retains any references to the test classes for the shortest
+    # possible time, as any remaining reference will prevent the class
+    # from being garbage collected.
+    #
+    # Test classes should be anonymous to support this. A class name
+    # is a constant reference to the class, preventing garbage
+    # collection unless the constant is redefined or undefined first
+    # (and maybe not even then!).
+
+    # Call the GC after each example to try to clean up the
+    # definitions of the test classes.
+    #
+    # In some circumstances, something may hold a reference to a test
+    # class long enough to save it from the garbage collector in the
+    # after block. Force the GC both before and after to be sure.
+    before { ObjectSpace.garbage_collect }
+    after { ObjectSpace.garbage_collect }
+    # A final GC call, as if to say that we really mean it!
+    after(:all) { ObjectSpace.garbage_collect }
```

Comment:
It depends on the Ruby version and, I suspect, on timing.

An object won't be cleaned up on GC if there's a live reference to it. Unfortunately, it's hard to predict how long a given Ruby version may hold on to a reference to a class. These difficulties multiply when you throw more versions into the mix.

For example, under some versions, these tests fail without the GC in the `before` block. In some versions, accessing `instance.class` seems to cause a longer-lived reference. In some versions, it appears that if you name the class, you may never get rid of it. I even had GC failures once when I compared a class to itself! That's why the assertions themselves are very careful about what properties they assert on.

Basically, the `before` and `after` blocks are both necessary in some Ruby versions. The `after(:all)` may or may not be needed. It _is_ a bit of prayer: it might not be needed and might not do anything but sometimes it might be answered when needed, and it shows my devotion to cleaning up after myself. It's a small, harmless devotion to propitiate the interpreter.

I take this all as a sign that trawling `ObjectSpace` to find the dependency groups is not the best approach.